// @spartan-ng/brain/fesm2022/spartan-ng-brain-forms.mjs
import * as i0 from '@angular/core';
import { Injectable, computed } from '@angular/core';
import { NgForm } from '@angular/forms';
import { toSignal } from '@angular/core/rxjs-interop';
import { BehaviorSubject, of } from 'rxjs';
import { switchMap, startWith, map, catchError } from 'rxjs/operators';

/**
 * Provides a custom {@link ErrorStateMatcher} at the component or application level.
 *
 * @example <caption>Application-level (appConfig)</caption>
 * ```ts
 * provideErrorStateMatcher(ShowOnDirtyErrorStateMatcher)
 * ```
 *
 * @example <caption>Component-level</caption>
 * ```ts
 * providers: [provideErrorStateMatcher(ShowOnDirtyErrorStateMatcher)]
 * ```
 */
function provideErrorStateMatcher(matcher) {
    return { provide: ErrorStateMatcher, useClass: matcher };
}
/** Error state matcher that matches when a control is invalid and dirty. */
class ShowOnDirtyErrorStateMatcher {
    isInvalid(control, form) {
        return !!(control && control.invalid && (control.dirty || (form instanceof NgForm && form.submitted)));
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ShowOnDirtyErrorStateMatcher, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    /** @nocollapse */ static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ShowOnDirtyErrorStateMatcher });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ShowOnDirtyErrorStateMatcher, decorators: [{
            type: Injectable
        }] });
/**
 * Determines when a form control should be considered in an invalid (error) state.
 *
 * The return value of `isInvalid` is reflected as the `data-matches-spartan-invalid` attribute
 * on field components, which drives both error styling and the visibility of `HlmFieldError`
 * messages.
 *
 * The default implementation matches when the control is invalid and either touched or, for
 * template-driven forms, the parent `NgForm` has been submitted.
 *
 * Provide a custom implementation (e.g. `ShowOnDirtyErrorStateMatcher`) at the component or
 * application level to change when errors are shown.
 */
class ErrorStateMatcher {
    isInvalid(control, form) {
        return !!(control && control.invalid && (control.touched || (form instanceof NgForm && form.submitted)));
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ErrorStateMatcher, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    /** @nocollapse */ static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ErrorStateMatcher, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ErrorStateMatcher, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });

class ErrorStateTracker {
    _matcher;
    _parentFormGroup;
    _parentForm;
    _ngControl = new BehaviorSubject(null);
    controlState = toSignal(this._ngControl.pipe(switchMap((ngControl) => {
        if (!ngControl)
            return of(null);
        const control = ngControl.control;
        if (!control) {
            throw new Error(`ErrorStateTracker: ngControl.control is null or undefined. This usually means the form control directive (e.g. formControlName or ngModel) ` +
                `found an NgControl in the DI tree but no ControlValueAccessor was registered to back it. ` +
                `Ensure the host component or directive provides NG_VALUE_ACCESSOR (e.g. { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(() => YourComponent), multi: true }) ` +
                `and implements the ControlValueAccessor interface.`);
        }
        return control.events.pipe(startWith(() => this._getState(control)), map(() => this._getState(control)));
    }), catchError((error) => {
        console.error(error);
        return of(null);
    })), {
        requireSync: true,
        equal: (a, b) => JSON.stringify(a) === JSON.stringify(b),
    });
    errors = computed(() => this.controlState()?.errors ?? null, ...(ngDevMode ? [{ debugName: "errors" }] : []));
    dirty = computed(() => this.controlState()?.dirty ?? null, ...(ngDevMode ? [{ debugName: "dirty" }] : []));
    invalid = computed(() => this.controlState()?.invalid ?? null, ...(ngDevMode ? [{ debugName: "invalid" }] : []));
    spartanInvalid = computed(() => this.controlState()?.spartanInvalid ?? null, ...(ngDevMode ? [{ debugName: "spartanInvalid" }] : []));
    touched = computed(() => this.controlState()?.touched ?? null, ...(ngDevMode ? [{ debugName: "touched" }] : []));
    get _controlParent() {
        return this._parentFormGroup || this._parentForm;
    }
    constructor(_matcher, _parentFormGroup, _parentForm) {
        this._matcher = _matcher;
        this._parentFormGroup = _parentFormGroup;
        this._parentForm = _parentForm;
    }
    setControl(ngControl) {
        // Wait for next tick so the ngControl.control property gets initialized
        Promise.resolve().then(() => {
            this._ngControl.next(ngControl);
        });
    }
    _getState(control) {
        const spartanInvalid = this._matcher?.isInvalid(control, this._controlParent) ?? false;
        return {
            dirty: control.dirty,
            errors: control.errors,
            invalid: control.invalid,
            spartanInvalid: spartanInvalid,
            touched: control.touched,
        };
    }
}

/**
 * Generated bundle index. Do not edit.
 */

export { ErrorStateMatcher, ErrorStateTracker, ShowOnDirtyErrorStateMatcher, provideErrorStateMatcher };
